Add the introduction part to Snippet Tutorial

Ninimama opened the pull request:
Corresponding to Issue #374 


# Pull Request Checklist

Below is a simple checklist but please do not hesitate to ask for assistance!

- [x] Fork, clone, and checkout the newest version of the code
- [x] Create a new branch
- [x] Make necessary code changes
- [x] Install `black` (i.e., `python -m pip install black` or `conda install -c conda-forge black`)
- [x] Install `flake8` (i.e., `python -m pip install flake8` or `conda install -c conda-forge flake8`)
- [x] Install `pytest-cov` (i.e., `python -m pip install pytest-cov` or `conda install -c conda-forge pytest-cov`)
- [x] Run `black .` in the root stumpy directory
- [x] Run `flake8 .` in the root stumpy directory
- [ ] Run `./setup.sh && ./test.sh` in the root stumpy directory
- [x] Reference a Github issue (and create one if one doesn't already exist)

Ninimama commented:
@seanlaw:
I didn't do `./setup.sh && ./test.sh` in the root stumpy directory as it takes a long time on my PC and I remembered you said it is not necessary when doing Tutorial. Right?

And, as you suggested, I tried to submit the first PR on this issue that includes only small changes. I will consider that 50 lines rule of thumb for subsequent pushes.

seanlaw commented on docs/Tutorial_Time_Series_Snippets.ipynb:
```diff
@@ -1,5 +1,29 @@
 {
```
<p>@ninimama I think this is pretty good and here are a couple of comments:</p><p> </p><ul><li>In your opening sentence, it is claimed that it is a "common question" but it would be nice to provide some (obvious) real-world examples if possible</li><li>I like the point that you are trying to make about snippets vs clustering but the clustering point could be a bit clearer</li><li>"Although a clustering" should be "Although clustering"</li><li>Since we acknowledge the original papers in the first sentence, there is no need to repeat things like "Authors in Matrix Profile XIII". We make no claims of these ideas being our own original ideas </li><li>So, I don't think the final tutorial will go through the creation of the toy data. We will just create it in Zenodo.ipynb and then download the final data to analyze</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/397/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Ninimama replied:
@seanlaw 

> * So, I don't think the final tutorial will go through the creation of the toy data. We will just create it in Zenodo.ipynb and then download the final data to analyze

In my recent commit, I keep the toy data for now and just add textual context and improve the current flow/figures. I tried to follow the same structure used in the other tutorials. 

I can take care of retrieving the data from Zenodo in the future commits (If you have any suggestion regarding this matter, please let me know) 

================================================

I have one question that has been bothering me:

Sometimes I try to limit the changes into one particular area to avoid the term "and" in the commit message. In other words, I would like to consider each commit for one particular issue. So, it would be easier for the reviewer (you) to go through the changes in the commit. However, in some occasions, there are only a few changes that need to be done for a particular issue. For instance, adding the heading to the notebook is probably less than 10 lines. Or, modifying figures (and change their orders) is again less than 10 lines. This can easily happen in different scenarios, when I try to improve different parts of the Tutorial. 

Is that okay to do one git commit for all those small changes? (This is what I did in my recent commit that I pushed) 

OR: Can I do three small commits (each corresponds to small changes). Then, after a few commits, I can do git push. Since I don't know how the review process looks like on your end, I have no idea if the latter approach makes sense or not.

seanlaw commented:
> I didn't do ./setup.sh && ./test.sh in the root stumpy directory as it takes a long time on my PC and I remembered you said it is not necessary when doing Tutorial. Right?

Yes, that should be fine

> And, as you suggested, I tried to submit the first PR on this issue that includes only small changes. I will consider that 50 lines rule of thumb for subsequent pushes.

Sounds good! Thanks

Ninimama commented:
@seanlaw 

> * So, I don't think the final tutorial will go through the creation of the toy data. We will just create it in Zenodo.ipynb and then download the final data to analyze

In my recent commit, I keep the toy data for now and just add textual context and improve the current flow/figures. I tried to follow the same structure used in the other tutorials. 

I can take care of retrieving the data from Zenodo in the future commits (If you have any suggestion regarding this matter, please let me know) 

================================================

I have one question that has been bothering me:

Sometimes I try to limit the changes into one particular area to avoid the term "and" in the commit message. In other words, I would like to consider each commit for one particular issue. So, it would be easier for the reviewer (you) to go through the changes in the commit. However, in some occasions, there are only a few changes that need to be done for a particular issue. For instance, adding the heading to the notebook is probably less than 10 lines. Or, modifying figures (and change their orders) is again less than 10 lines. This can easily happen in different scenarios, when I try to improve different parts of the Tutorial. 

Is that okay to do one git commit for all those small changes? (This is what I did in my recent commit that I pushed) 

OR: Can I do three small commits (each corresponds to small changes). Then, after a few commits, I can do git push. Since I don't know how the review process looks like on your end, I have no idea if the latter approach makes sense or not.

seanlaw commented:
@Ninimama I can handle the Zenodo upload part. Let's just keep it as is for now

Regarding commits, I wouldn't worry so much about tutorials. I think you can bundle a few things into a single commit for a notebook and then let me know in this comments section if there  is anything that I need to know.

I can tell you that I am not looking at every single commit. In fact, I'm only looking at the latest version via the "ReviewNB" purple button above.

seanlaw commented on docs/Tutorial_Time_Series_Snippets.ipynb:
```diff
@@ -1,8 +1,51 @@
 {
```
<p>I don't think you need a header called "Snippets". </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/397/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

seanlaw replied:
<ul><li>You said "common question" but you've only presented a statement. </li><li>Please remove "large" in "large time series" and then combine that sentence with the one following it</li><li>Please replace "his sleep" with "their sleep"</li></ul><p> </p><p>This paragraph seems somewhat forced and out of place:</p><p> </p><blockquote>For instance, the respiration data of a person recorded during his sleep can be reported by showing only the top-2 snippets. As an another example, snippets can be used to find different stages in the walking-nordic walking-skipping-running cycle of a person. Particularly, the y-axis acceleration data recorded by a chest-worn sensor can be summarized into top-4 snippets where each snippet shows a different type of exercise.  </blockquote><p> </p><p>This may be something that we'll need to think through and provide proper motivation without getting into too much unnecessary detail. </p><p> </p><p>I think that the <code>It is worthwhile to note what snippet is NOT </code>  part should not be in the starting section and should be its own section header</p><p> </p><p>This sentence isn't so clear:</p><p> </p><blockquote>one should specify the beginning and ending of subsequences and then apply clustering on the set of subseuqences</blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/397/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Ninimama commented:
@seanlaw 

> This may be something that we'll need to think through and provide proper motivation without getting into too much unnecessary detail.

I am still thinking about it (I put "?" in the corresponding part in the notebook.

===================================================

Just to confirm: Would you like to reproduce the result of another data (other than the toy data currently provided in the notebook)?

seanlaw commented:
> Just to confirm: Would you like to reproduce the result of another data (other than the toy data currently provided in the notebook)?

I like the toy data as it is simple/easy to understand and it helps to lay the foundation for explaining things. However, I am open to adding a slightly more interesting example from the paper (if it exists) AFTER the toy data set. Of course, this runs the risk of making the tutorial too long/redundant. Unfortunately, it's one of those things that we'll need to decide whether it makes the tutorial "better"/"worse" only after it is added. How does that sound?

Ninimama commented:
> . Of course, this runs the risk of making the tutorial too long/redundant.

Exactly! In fact, I, as a reader, prefer to understand the basics quickly and easily, and then go and explore my own data. 

==================================================================

The reason I asked was that I misinterpreted one of our previous conversations and I thought we should add a real data set. So, for now, I can try to polish the tutorial and see if I can find a good example of snippets application.

Ninimama commented on docs/Tutorial_Time_Series_Snippets.ipynb:
```diff
@@ -1,8 +1,84 @@
 {
```
@seanlaw 
I have been busy with my research lately. Will take care of it and the other things soon.

seanlaw replied:
> I have been busy with my research lately. Will take care of it and the other things soon.

@Ninimama There is absolutely no rush. Quality over quantity! Please take your time

Ninimama commented:
@seanlaw 

According to the github guide, I did `git fetch upstream` to synch my fork (on github) with the origin repo  but it shows the following error:

> fatal: couldn't find remote ref refs/heads/master

According to a stackoverflow question, I checked out the .git/config file (shown below) but things sound okay to me:

![image](https://user-images.githubusercontent.com/38519522/122623326-d984f180-d058-11eb-914c-d8cdbc8b427a.png)


I also tried `git fetch` and `git fetch origin`. But, none of them works. 

Could you please help me resolve this issue? (So, I can import mplstyle in my local repo)

seanlaw commented:
@Ninimama I'm not entirely sure but it may be due to the fact that `refs/heads/master` does not exist in the `TDAmeritrade/stumpy.git` repo since we actually use `refs/heads/main`.  I recommend following this documentation:

https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

Ninimama commented:
@seanlaw 
I did follow that before. In fact, `git remote -v` shows the upstreams. But, I just removed them and added them again. 

And, now it works! 
Thanks

seanlaw commented:
> I did follow that before. In fact, git remote -v shows the upstreams. But, I just removed them and added them again.

After asking around, I also learned that there is a button in your fork:

![E4M-4IYXoAwDUUA](https://user-images.githubusercontent.com/7473521/122624834-0d1f4600-d070-11eb-9eb7-99bb3c2e023d.jpeg)

seanlaw commented:
@Ninimama Did you also happen to follow step 5 in https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork?  It is one step to sync your fork with the `TDAmeritrade/stumpy.git` repo but it is another step to sync your branch with your fork's main. 

I think this important process is missing from our contributor's guide. Would you mind helping us add it if you have the time? No rush.

Ninimama commented:
@seanlaw 

I think I did.  And, if I remember correctly, the stumpy.mplstyle  appears in the main branch (the default local branch on my laptop); however, the same file doesn't appear in the branch Snippet_Tutorial. 

So, I did `git switch `Snippet_Tutorial && git rebase main` and now it is okay.

Ninimama commented:
@seanlaw 
I checked out @alvii147 PR and I think he already updated the contributing doc. 

As I said, I should have first checked out the Development branch first and then do the merge. 

NOTE:
I am not sure if `git rebase` is better than `git merge` or not.

seanlaw commented:
@Ninimama I believe that both will work. However, `git merge main` is probably preferred as it preserves history. I came across this article that explains it quite well:

https://www.atlassian.com/git/tutorials/merging-vs-rebasing

Ninimama commented:
@seanlaw 

Could you please help me out to figure the `git push` issue below:
FYI: I did `git fetch upstream` before starting to work on my files today, but I forgot to do `git merge`. Then, I finished my work while I was at `Snippets_Regime` branch (a sub-branch of `Snippets_Tutorial`). Then, I merged it to its parent branch, `Snippets_Tutorial`. I got error for `git push origin Snippets_Tutorial` (the same error to the ones you can see in the following images). I tried to resolve it. Then, I realized I should have done `git merge upstream/main` before I started working on those files.  I did that, but then I was asked to commit new/modified files of remote repo  to allow git merge. But again, I still get the error in git push. 

Any opinion? (I can remove the branch and submit a new PR if there is no other way around this)

![image](https://user-images.githubusercontent.com/38519522/123054766-6fef4500-d3c2-11eb-8e79-c1716fc0d9a7.png)

![image](https://user-images.githubusercontent.com/38519522/123054840-84334200-d3c2-11eb-9285-59261b35d600.png)

Sorry for the headache!

Ninimama commented:
@seanlaw 
UPDATE:

Now, I tried the `fetch upstream` of the GitHub and the follow the instructions again. But, the same error:

![image](https://user-images.githubusercontent.com/38519522/123068414-1c372880-d3cf-11eb-9d67-93b01a0485b4.png)

I can also see an error that says `Snippets_Tutorial has no upstream branch`. Isn't it weird? (as I set that `upstream` branch before)

=========================================================

I also checked out this link: [atlassian: git-merge](https://www.atlassian.com/git/tutorials/using-branches/git-merge)

If I understand correctly, this case is probably a 3-way merge. And, according to the provided link above, it might be better to rebase my new-feature branch onto the main branch if the changes of new-feature is small.

alvii147 commented:
@Ninimama try doing `git pull` before you do `git push` in the `Snippets_Tutorial` branch. The error message is telling you your remote branch is ahead of your local branch and has changes that are not present locally.

seanlaw commented:
> @Ninimama try doing `git pull` before you do `git push` in the `Snippets_Tutorial` branch. The error message is telling you your remote branch is ahead of your local branch and has changes that are not present locally.

I was going to recommend the same thing

Ninimama commented:
@seanlaw  @alvii147 

Thanks for the response. So, I tried:
`git pull origin main` while I was at my local branch, `Snippets_Tutorial`. It says things are good. So, I tried then:
`git pull origin Snippets_Tutorial`. And, it says I should resolve the conflicts first. Any idea how to do so as I cannot open the ipynb file in jupyter:

![image](https://user-images.githubusercontent.com/38519522/123132341-e06e8400-d40b-11eb-8044-fa63095b86d1.png)

alvii147 commented:
@Ninimama looking at the error message in the screenshot you provided, the `<<<<<<<<HEAD` is just git's way of pointing to conflicts. Open up `docs/Tutorial_Time_Series_Snippets.ipynb` and search for that line, that's where the conflict is. Resolve the conflict in that line and then try pulling again.

[This](https://www.youtube.com/watch?v=__cR7uPBOIk) is a good tutorial for resolving merge conflicts.

seanlaw commented:
> I installed VS code and it can show the script. However, it is not really better than notepad++. One way is to abort the merge and copy my notebook somewhere else (because the conflict in the branch is only caused by docs/Tutorial_Time_Series_Snippets.ipynb. Then, do the git pull again (I assume there should be no conflict then because I completely deleted the docs/Tutorial_Time_Series_Snippets.ipynb. Then, open them and manually resolve conflict. The drawback is that I will not see the marker and I need to see them via VS code.

So, merge conflicts means that there were changes to the tutorial elsewhere and then you changed your local version within your branch and Git could not resolve the conflict (i.e., it usually requires a human to decide which version and lines of code to keep). Since it's a Tutorial, what you did was perfectly fine (i.e., copy the "right" version somewhere else, checkout the tutorial from upstream, and then overwrite the tutorial with the "right" version afterward).

seanlaw commented:
@alvii147 That might be a good resource to add to the end of the contributing guide as "additional resources"

alvii147 commented:
@seanlaw that's a good idea, merge conflicts can be nightmares sometimes. I'm gonna look into doing that once the annotation vectors tutorial is off my plate.

Ninimama commented:
@seanlaw 
I FINALLY handle the issue. Please ignore most of the recent commits, as I was trying to handle the merge conflict. In the last commit, I committed the changes to the notebook. Furthermore, you should be able to see the modified version of `snippets.py` and `aampdist_snippets.py` modules as well (I added the regimes). Any opinion on the next step? Writing code for the unit test of the regimes?

seanlaw commented:
Please give me some time to review

seanlaw commented on stumpy/aampdist_snippets.py:
```diff
@@ -8,6 +8,30 @@
+    (starts, ) = np.where(~m1 & m2)
+    (ends, ) = np.where(m1 & ~m2)
+
+    slices = np.c_[[idx] * len(starts), starts, ends]
```
Can you please explain what is happening here in this line? The original code had:

```
def _get_mask_slices(mask):
    m1 = np.r_[0, mask]
    m2 = np.r_[mask, 0]
    starts, = np.where(~m1 & m2)
    ends, = np.where(m1 & ~m2)
    return np.c_[starts, ends]
```

and this looks different

Ninimama replied:
I wanted to provide the snippet ID as well when I return the slices. (However, as you said, it is better to use the reusable version and change the output to include the snippet ID).

seanlaw replied:
Ahhh, now I understand. Thank you for clarifying it for me!

seanlaw commented on stumpy/aampdist_snippets.py:
```diff
@@ -213,6 +237,8 @@ def aampdist_snippets(
     Q = np.full(D.shape[-1], np.inf)
     indices = np.arange(0, n_padded - m, m)
+    snippets_regimes = []
+    snippets_regimes = np.empty((0, 3), int)
```
Why are we creating an empty list and then overwriting it with an empty array?

We should just stick to using an empty list and append arrays to the list and then once it is filled then we can convert it to an array all at once instead of appending to the array (as that is much slower)

Ninimama replied:
> Why are we creating an empty list and then overwriting it with an empty array?
My mistake

> We should just stick to using an empty list and append arrays to the list and then once it is filled then we can convert it to an array all at once instead of appending to the array (as that is much slower)

Didn't know that.

seanlaw commented on stumpy/aampdist_snippets.py:
```diff
@@ -231,11 +257,14 @@ def aampdist_snippets(
         snippets_fractions[i] = np.sum(mask) / total_min.shape[0]
         total_min = total_min - mask.astype(np.float64)
+        slices = _get_mask_slices(i, mask)
+        snippets_regimes = np.append(snippets_regimes, slices, axis=0)
```
Assuming you start with an empty Python list (see above) then you'd do `snippets_regimes.append(slices)` here 

and then, afterward, you'd convert this list to an array:

```
snippets_regimes = np.array(snippets_regimes)
```

seanlaw commented on stumpy/snippets.py:
```diff
@@ -10,6 +10,30 @@
 from .aampdist_snippets import aampdist_snippets
 
 
+def _get_mask_slices(idx, mask):
```
Considering that this function is identical to the one in aampdist_snippets.py, I think this function should be moved into `core.py`

seanlaw commented on stumpy/snippets.py:
```diff
@@ -222,6 +246,7 @@ def snippets(
     snippets_areas = np.empty(k)
     Q = np.full(D.shape[-1], np.inf)
     indices = np.arange(0, n_padded - m, m)
+    snippets_regimes = np.empty((0, 3), int)
```
Please see comments above

seanlaw commented on stumpy/snippets.py:
```diff
@@ -240,11 +265,14 @@ def snippets(
         snippets_fractions[i] = np.sum(mask) / total_min.shape[0]
         total_min = total_min - mask.astype(np.float64)
+        slices = _get_mask_slices(i, mask)
+        snippets_regimes = np.append(snippets_regimes, slices, axis=0)
```
Please see comments above

seanlaw commented on stumpy/snippets.py:
```diff
@@ -10,6 +10,30 @@
+    (starts, ) = np.where(~m1 & m2)
+    (ends, ) = np.where(m1 & ~m2)
+
+    slices = np.c_[[idx] * len(starts), starts, ends]
```
I think this function with `idx` is too specialized/specific. I think it is better to stick with the more generic (and reusuable):

```
def _get_mask_slices(mask):
    m1 = np.r_[0, mask]
    m2 = np.r_[mask, 0]
    starts, = np.where(~m1 & m2)
    ends, = np.where(m1 & ~m2)
    return np.c_[starts, ends]
```

And then post-process the array once it is returned. Otherwise, it isn't clear what `idx` is doing

seanlaw reviewed (changes requested):
@Ninimama Here are some suggestions for you to consider.

Ninimama commented:
@seanlaw 
I revised the modules according to your suggestion. After I move the function `_get_mask_slices(mask)` to the module `core.py`, I added the line `from .core import ...` in the beginning of the `snippets.py` and `aampdist_snippets.py` modules. 

However, the line `from .core import ...` already existed: 
`from .core import check_window_size`

So, I just checked out the [recommended style](https://www.python.org/dev/peps/pep-0008/) and I realized I should have better put them together in single line and use comma to separate them.

I can modify it next time

========================================

Should I now think of writing the unit tests for these modules?

seanlaw commented:
> To provide the snippet ID in the slices, I used the np.insert(). If it is slow as it tries to create a new array by inserting the new value, is there any faster way to include the snippet ID in the slices returned by the function `_get_mask_slices(mask)`?
> Just to confirm, we finally return an array of arrays. This is what the `snippets.py` and `aampdist_snippets.py` return in their current version in the `Snippet_Tutorial` branch

So, I'm thinking about something like:

```
# core.py 
def _get_mask_slices(mask):
    m1 = np.r_[0, mask]
    m2 = np.r_[mask, 0]
    starts, = np.where(~m1 & m2)
    ends, = np.where(m1 & ~m2)
    return np.c_[starts, ends]

# snippets.py or aampdist_snippets.py
from .core import check_window_size, _get_mask_slices
...
snippets_regimes_list = []
for i in range(k):
        profile_areas = np.sum(np.minimum(D, Q), axis=1)
        mask = snippets_profiles[i] <= total_min
        snippets_fractions[i] = np.sum(mask) / total_min.shape[0]
        total_min = total_min - mask.astype(np.float64)
        slices = _get_mask_slices(mask)
        snippets_regimes_list.append(slices)

snippets_regimes = np.empty((len(snippets_regimes_list), 3), dtype=object)
snippets_regimes[:, 0] = np.repeat(np.arange(len(snippets_regimes_list)), [regime.shape[0] for regime in snippets_regimes_list])
snippets_regimes[:, 1:] = np.vstack(snippets_regimes_list)
```

I haven't tested this yet so it may require some adjustment but hopefully you get the point. 

> I revised the modules according to your suggestion. After I move the function `_get_mask_slices(mask)` to the module `core.py`, I added the line `from .core import ...` in the beginning of the `snippets.py` and `aampdist_snippets.py` modules.
> 
> However, the line `from .core import ...` already existed:
> `from .core import check_window_size`
> 
> So, I just checked out the [recommended style](https://www.python.org/dev/peps/pep-0008/) and I realized I should have better put them together in single line and use comma to separate them.
> 
> I can modify it next time

Yes, please combine the imports into a single line

> ========================================
> 
> Should I now think of writing the unit tests for these modules?

Yes. I think you should start with writing the unit test for `_get_mask_slices` first. If you look inside `tests/test_core.py` you'll see tests for each function from `core.py`. I would like to see a naive implementation of `_get_mask_slices` using simple Python `for-loops`

Ninimama commented:
> > Should I now think of writing the unit tests for these modules?
> 
> Yes. I think you should start with writing the unit test for `_get_mask_slices` first. If you look inside `tests/test_core.py` you'll see tests for each function from `core.py`. I would like to see a naive implementation of `_get_mask_slices` using simple Python `for-loops`


I finished the unit tests. But before commit and push it. I would like to know if you have any particular definition of the concept of native code. For instance, below is my implementation:

```
def naive_get_mask_slices(mask):
    if np.sum(mask)==0:
      return np.empty(shape=(0,2), dtype="int64")

    else:
        slices_list = []
        prev_val = False
        
        for i in range(len(mask)):
            if mask[i] and ~prev_val and (i < len(mask)-1): 
                #finding the start index of a new slice
                new_slice = [i]
                prev_val = mask[i]
            elif mask[i] and ~prev_val and (i ==len(mask)-1): 
                #finding the start index of a new slice, mask is exhausted
                slices_list.append([i,i+1])
            elif ~mask[i] and prev_val:
                #finding the end index of the current slice
                new_slice.append(i)
                slices_list.append(new_slice)
                new_slice = []
                prev_val = mask[i]
            elif (mask[i] and prev_val and (i==len(mask)-1)):
                #finding the end index of the current slice, mask is exhausted
                new_slice.append(i+1)
                slices_list.append(new_slice)
            else:
                continue
``` 
**I also wrote a test function to compare it with the output of `_get_mask_slices`. And, it works** 

===================================================

However, after a quick search, I found a piece of code that is used in `more_itertools` package. And, I just modified it for our purpose:

```
bool_arr = np.array([True, True, True, False, False, True, True, False, False, True])

def consecutive_groups(iterable, ordering=lambda x: x):
  from itertools import groupby
  from operator import itemgetter
  for k, g in groupby(
    enumerate(iterable), key=lambda x: x[0] - ordering(x[1])
    ):
    yield map(itemgetter(1), g)
  
all_slices=[]
for group in consecutive_groups(np.where(bool_arr)[0]):
    lst = list(group)
    all_slices.append([lst[0], lst[-1]+1])

np.array(all_slices)

# output:
array([[ 0,  3],
       [ 5,  7],
       [ 9, 10]])
```

And, both `itertools` and `operator` are python standard libraries. Do you think the latter piece of code is better?

seanlaw commented:
> However, after a quick search, I found a piece of code that is used in more_itertools package. And, I just used it for our purpose as follows:

So, we strongly prefer not adding any additional dependencies for testing as it affects our ability to maintain it long term. Self contained code is always preferred. I think your function is fine though the many `elif` statements seems to be a red-flag to me. I'm guessing that there might be a way to simplify this.

seanlaw commented:
Can you check if something like this might suffice?

```
def naive_get_mask_slices(mask):
    tmp = np.r_[0, mask]
    idx = []
    for i, val in enumerate(np.diff(tmp)):
        if val == 1:
            idx.append(i)
        if val == -1:
            idx.append(i)

    if tmp[-1]:
        idx.append(len(mask))

    return np.array(idx).reshape(len(idx)//2, 2)
```

Ninimama commented:
> Can you check if something like this might suffice?

your implementation of the function `naive_get_mask_slices` worked ! 


> I haven't tested this yet so it may require some adjustment

Regarding the `snippets_regimes`, I just did a minor adjustment to correct the output.

seanlaw commented on stumpy/core.py:
```diff
@@ -1806,3 +1806,31 @@ def _jagged_list_to_array(a, fill_value, dtype):
+    Parameters
+    ----------
+    mask: ndarray
+          A boolean 1D array
```
The indentation looks off here

seanlaw commented on stumpy/core.py:
```diff
@@ -1806,3 +1806,31 @@ def _jagged_list_to_array(a, fill_value, dtype):
+    Returns
+    -------
+    slices: ndarray
+            slices of indices where the mask is True. Each slice has a size of two:
```
The indentation looks off here

seanlaw commented on tests/test_core.py:
```diff
@@ -827,3 +827,34 @@ def test_jagged_list_to_array_empty():
     npt.assert_array_equal(left, right)
+
+
+def naive_get_mask_slices(mask):
```
Please move this naive implementation to the top of the test file. See other test files to see the location of the naive implementation

Ninimama replied:
> Please move this naive implementation to the top of the test file. See other test files to see the location of the naive implementation

I moved it to the TOP of the `test_core.py` file. 
FYI: Just in case you missed it, NOT all naive implementations are at the TOP. there are some `test` functions between the `naive` implementations.

seanlaw replied:
Okay, would you mind creating a new issue to track this? I think it's best when the naive implementations are at the top of the file and then followed by the tests below it

seanlaw commented on tests/test_core.py:
```diff
@@ -827,3 +827,34 @@ def test_jagged_list_to_array_empty():
+
+
+def test_get_mask_slices():
+    bool_lst = [False, True]
```
Cool

Ninimama commented:
@seanlaw 
I will take a look at the `test_snippets.py` and `test_aampdist_snippets.py` to include the test for the `snippets_regime`. 

I noticed you used the naive implementation of mpdist in the unit test of snippets. I was wondering if I should move the function `naive_get_mask_slices` to the `naive.py` file and reuse it again later in the unit test of snippets modules.

seanlaw commented:
> I will take a look at the test_snippets.py and test_aampdist_snippets.py to include the test for the snippets_regime.

> I noticed you used the naive implementation of mpdist in the unit test of snippets. I was wondering if I should move the function naive_get_mask_slices to the naive.py file and reuse it again later in the unit test of snippets modules.

Yes, if you plan to use `naive_get_mask_slices` in one than one `test` file then please move `naive_get_mask_slices` into `naive.py`. `naive.py` is there to help us avoid writing the same piece of code in multiple places.

seanlaw commented:
@Ninimama Nit trying to rush you as I'm sure you are busy but I just wanted to check in in case you need something from me. Please let me know how I can help

Ninimama commented:
@seanlaw 

> @Ninimama Nit trying to rush you as I'm sure you are busy but I just wanted to check in in case you need something from me. Please let me know how I can help

Thanks for checking in. I took a look but then I got a little busy doing my research. I will try to finish this at the end of this week.

Ninimama commented:
@seanlaw 

I revised the `test_snippets.py` and `test_aampdist_snippets.py` to include regimes in the unit test. I also revise the `naive.mpdist_snippets.py` and `naive.aampdist_snippets.py` to include the regimes. However, I still get error and cannot figure it out why. I took a look at the ./test.sh file and it seems you use -rx option which should show the details when a test fails. But, this is what I get when I run `./test.sh`

![image](https://user-images.githubusercontent.com/38519522/126091694-959cd085-64ed-4875-afb3-d29320e09aca.png)

seanlaw commented:
Yes, please push your latest changes and I can take a look.

seanlaw commented:
@Ninimama So, it looks like the first thing that needs to be fixed is the code formatting. For STUMPY, we use the [`black` code formatter](https://black.readthedocs.io/en/stable/) and I can see that it has already been installed on your system as evidenced by the earlier error:

```
Oh no! 💥 💔 💥
1 file would be reformatted, 74 files would be left unchanged.
Error: Process completed with exit code 1.
``` 

By running `cd contributions\stumpy && black .`, this will tell black to automatically correct the formatting of your code if it finds any inconsistencies that it can resolve. Note that `black` does not change the actual logic of the code. It just makes sure that things like indentation, spacing, and line length are the same across the entire codebase but this ensures that everything looks the same aesthetically. 

After you've resolved this, please commit the changes and push again.

Ninimama commented:
@seanlaw 

> we use the [`black` code formatter](https://black.readthedocs.io/en/stable/) and I can see that it has already been installed on your system

I thought black can automatically fix things and save them. But, after your note, I realized I have to do it and then do `git add/ commit`. I also noticed that I previously didn't use `.` in `black .` 
==========================================================

There we go! Finally! :) Thanks for your input.

==========================================================

**Just one note:** 

The intro part of the tutorial notebook for `snippet` needs to be revised. It still misses a good real-world yet easily understandable example for the use of the snippet algorithm.

seanlaw commented:
> I thought black can automatically fix things and save them. But, after your note, I realized I have to do it and then do `git add/ commit`. I also noticed that I previously didn't use `.` in `black .` 

As mentioned in the [Contributing guide](https://stumpy.readthedocs.io/en/latest/Contribute.html#Adhere-to-CONTRIBUTING.md-Guidance), for Python code, you should run both `black .` and then `flake8 .`

> The intro part of the tutorial notebook for `snippet` needs to be revised. It still misses a good real-world yet easily understandable example for the use of the snippet algorithm.

Do you have any suggestions for this? I think the current example is simple and easy to understand but I agree that it would be nice to have a "real" example

Ninimama commented:
> Do you have any suggestions for this? I think the current example is simple and easy to understand but I agree that it would be nice to have a "real" example

I, unfortunately, don't have any nice examples in mind. The authors provided several examples in the paper; however, I think it is not completely easy to understand them. 

One of the examples was about the electricity demand and it is somehow understandable why the electricity demand is different in summer and winter (two snippets). Since I am an electrical engineer, I might be a little biased. 

I think we can leave it as it is. In the future, some of the users might come forward and share their stories and provide a nice example for the application of this algorithm.

seanlaw commented:
> I think we can leave it as it is. In the future, some of the users might come forward and share their stories and provide a nice example for the application of this algorithm.

Sounds good. This PR is focused on adding the Introduction so we'll continue to work on this tutorial until it is ready. The unit tests look great! The naive implementation isn't "naive" enough for my liking (i.e., it is too similar to the original code) and so I'm going to adjust it after the PR is merged. Thanks again for working on this and please look out for the next release v1.9.0 coming out really soon!

I really appreciate all of your contributions @Ninimama and I have enjoyed collaborating with you!

Ninimama commented:
@seanlaw 
I really learned a lot. That was a really good experience and thanks for your support. Although I haven't used MP-based algorithms (and STUMPY) in my own research as I realized it might not be the proper tool for the purpose of my research, I enjoyed the algorithms and their application, and that's why I really enjoyed contributing to STUMPY package in the first place. And then, it was your support and how well you tried to keep things organized. Really Great Work!!!

seanlaw commented:
@Ninimama I made some changes/corrections to some of these files in [this commit](https://github.com/TDAmeritrade/stumpy/commit/6218b763a0783cff8b562c7b83bcccc93f924064)
